src/error.rs

```diff
@@ -95,10 +103,10 @@ impl fmt::Display for Error {
 
 #[derive(Debug)]
 pub enum ErrorKind {
-    MismatchedItemKind(Kind, Kind),
+    MismatchedItemKind(Item, Item),
```

Comment:
I thought it might be a bit cleaner / easier to piece together if we constructed each error with `Item`s rather than a `Kind` and `Guid` for each. However, this does inflate the size a bit: 
`size_of::<Item>() = 40 bytes`
`size_of::<Kind>() + size_of::<Guid>() = 25 bytes`

So it'd be 1.6x bigger, but I'm not entirely sure of the constraints in the system. 

If it's unnecessary, I can switch to `__(Kind, Item, Kind Item)`, `__((Kind, Item), (Kind, Item))`, or maybe change the error variants into structs so we have named fields to disambiguate.

Reply:
I don't think we're that concerned about the size here. If you're worried, making Error wrap `Box<ErrorKind>` instead would avoid problems and probably avoid a bunch of pointless memcpying in the success case as-is. In https://github.com/mozilla/application-services/ we usually box the error type for this reason, more or less.

That said, I doubt this is measurable compared to the overhead of the IO that dogear-using code tends to do, though (talking to a database and all).

Reply:
Yep, I also don't think we need to worry too much about the size increase. @thomcc is totally right, all the SQLite operations that we do in a-s and Firefox Desktop—and all the layers of wrapping—are going to have a much bigger impact than a dozen bytes. I definitely agree that it's much cleaner to piece the errors together like this, instead of unpacking into fields or tuples, too. 😄 Thanks for pointing it out!

Reply:
Makes total sense, I had figured that'd be the case but I just wanted to double check :smile: 

Thanks for clarifying!